### PR TITLE
feat: HIP-1215 Define the java function selectors needed in schedule contract translator

### PR DESCRIPTION
**Pull request by gkozyryatskyy**
**Description**:
For [HIP-1215](https://github.com/hiero-ledger/hiero-improvement-proposals/blob/main/HIP/hip-1215.md) implementation we need to define function selectors, translators and calls.

**Related issue(s)**:

Fixes [#20054](https://github.com/hiero-ledger/hiero-consensus-node/issues/20054)

**Notes for reviewer**:
- single `systemContract.scheduleService.scheduleCall.enabled` config is used for all `scheduleCall*` functions
- configs are `disabled` by default, while the feature is not ready
- `HasScheduleCapacityCallStub` is used for `hasScheduleCapacity(uint256,uint256)` stubbing while it is not implemented. Stub is not covered by unit tests because it doing nothing and will be removed.

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/ContractsConfig.java [hunk 1]
```diff
@@ -65,6 +65,15 @@ public record ContractsConfig(
         @ConfigProperty(value = "systemContract.scheduleService.scheduleNative.enabled", defaultValue = "true")
                 @NetworkProperty
                 boolean systemContractScheduleNativeEnabled,
+        @ConfigProperty(value = "systemContract.scheduleService.scheduleCall.enabled", defaultValue = "false")
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/ContractsConfig.java:68]: I believe we originally agreed to have a single flag for the whole feature, has that changed?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java [hunk 2]
```diff
@@ -164,7 +164,18 @@ public enum DispatchType {
...
+     */
+    SCHEDULE_CREATE_CONTRACT_CALL(HederaFunctionality.SCHEDULE_CREATE, SubType.SCHEDULE_CREATE_CONTRACT_CALL),
+    /**
+     * Dispatch for Hedera schedule delete functionality with default resource prices.
+     */
+    SCHEDULE_DELETE(HederaFunctionality.SCHEDULE_DELETE, DEFAULT),
+    /**
+     * Dispatch for Hedera schedule get info functionality with default resource prices.
+     * Currently decided to use 'gasCalculator.viewGasRequirement()' for all view functions, so this one is not used.
+     */
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java:178]: The java doc of `DispatchType` seems to be outdated:
> Enumerates the types of child transactions that can be dispatched to the HTS system contract.

-- it's not just HTS anymore.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java [hunk 3]
```diff
@@ -164,7 +164,18 @@ public enum DispatchType {
...
+     * Dispatch for Hedera schedule create functionality with schedule create contract call resource prices.
+     */
+    SCHEDULE_CREATE_CONTRACT_CALL(HederaFunctionality.SCHEDULE_CREATE, SubType.SCHEDULE_CREATE_CONTRACT_CALL),
+    /**
+     * Dispatch for Hedera schedule delete functionality with default resource prices.
+     */
+    SCHEDULE_DELETE(HederaFunctionality.SCHEDULE_DELETE, DEFAULT),
+    /**
+     * Dispatch for Hedera schedule get info functionality with default resource prices.
+     * Currently decided to use 'gasCalculator.viewGasRequirement()' for all view functions, so this one is not used.
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java:177]: `viewGasRequirement` seems to be using `DispatchType.TOKEN_INFO` 🤔 

if all view operations share the same dispatch type, should we name it accordingly? e.g. `DispatchType.ANY_VIEW_OP` or something similar?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/DispatchForResponseCodeHssCall.java [hunk 4]
```diff
@@ -139,6 +126,18 @@ public DispatchForResponseCodeHssCall(
...
+    /**
+     * Encodes the given {@code ContractCallStreamBuilder.status} and {@code ContractCallStreamBuilder.scheduleID}
+     * as a return value for a schedule create call.
+     *
+     * @param recordBuilder the result of the dispatch to encode
+     * @return the encoded status
+     */
+    public static ByteBuffer scheduleCreateResultEncode(@NonNull final ContractCallStreamBuilder recordBuilder) {
+        return RC_AND_ADDRESS_ENCODER.encode(
+                Tuple.of((long) recordBuilder.status().protoOrdinal(), headlongAddressOf(recordBuilder.scheduleID())));
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/DispatchForResponseCodeHssCall.java:145]: A general question, not really related to this PR:
Do you know if schedules share address space with other entities (e.g. account IDs or topic IDs), or each has its own? In other words, can there simultaneously exist an account 0.0.123 and a schedule 0.0.123? I expect it's shared, but I'm not sure.

Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/DispatchForResponseCodeHssCall.java:145]: actually, nvm, I'll ask that elsewhere 😄

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import javax.inject.Singleton;
+
+/**
+ * Translates {@code deleteSchedule*} calls to the HSS system contract. For details
+ * {@see <a href=https://github.com/hiero-ledger/hiero-improvement-proposals/blob/main/HIP/hip-1215.md>HIP-1215</a>}.
+ */
+@Singleton
+public class DeleteScheduleTranslator extends AbstractCallTranslator<HssCallAttempt> {
+
+    public static final SystemContractMethod DELETE_SCHEDULED = SystemContractMethod.declare(
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: I think the terminology we're going for is "delete schedule" as in "delete a schedule object", not "delete scheduled" as in "delete a scheduled call". So this const should be `DELETE_SCHEDULE` (no "D")? Same for `DELETE_SCHEDULE_PROXY` below.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+                    "deleteSchedule(address)", ReturnTypes.INT_64)
+            .withCategories(Category.SCHEDULE);
+    private static final int SCHEDULE_ID_INDEX = 0;
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: nitpicking: at the call level it's still a schedule _address_, so perhaps `SCHEDULE_ADDRESS_INDEX`?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+     * @param attempt the call attempt
+     * @return the schedule ID
+     */
+    @VisibleForTesting
+    public ScheduleID scheduleIdFor(@NonNull HssCallAttempt attempt) {
+        requireNonNull(attempt);
+        if (attempt.isSelector(DELETE_SCHEDULED)) {
+            final var call = DELETE_SCHEDULED.decodeCall(attempt.inputBytes());
+            final Address scheduleAddress = call.get(SCHEDULE_ID_INDEX);
+            final var number = ConversionUtils.numberOfLongZero(explicitFromHeadlong(scheduleAddress));
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: Do you think it would be worth adding an extra validation to check if the caller actually provides a long zero address (using `ConversionUtils.isLongZero`?).
Otherwise there are possibly many addresses that they could pass which would resolve to the same schedule ID (i.e. they could use an arbitrary prefix). I don't think that's a real issue, but it seems a bit cleaner if we handle that.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/getscheduledinfo/GetScheduledNonFungibleTokenCreateCall.java [hunk 8]
```diff
@@ -43,19 +43,24 @@ public PricedResult execute(MessageFrame frame) {
         final var schedule = nativeOperations().getSchedule(scheduleID);
         // Validate that given schedule exists
         if (schedule == null) {
-            return gasOnly(revertResult(RECORD_NOT_FOUND, gasCalculator.viewGasRequirement()), RECORD_NOT_FOUND, true);
+            return gasOnly(
+                    revertResult(RECORD_NOT_FOUND, gasCalculator.viewGasRequirement()), RECORD_NOT_FOUND, isViewCall);
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/getscheduledinfo/GetScheduledNonFungibleTokenCreateCall.java:47]: I appreciate cleaning those up 👍

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/hasschedulecapacity/HasScheduleCapacityTranslator.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import com.hedera.node.app.service.contract.impl.exec.utils.SystemContractMethod.Category;
+import com.hedera.node.app.service.contract.impl.exec.utils.SystemContractMethodRegistry;
+import com.hedera.node.config.data.ContractsConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Translates {@code deleteSchedule(address)} calls to the HSS system contract. For details
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/hasschedulecapacity/HasScheduleCapacityTranslator.java]: This one translates `hasScheduleCapacity` calls :)

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+import javax.inject.Singleton;
+
+/**
+ * Translates {@code scheduleCall*} calls to the HSS system contract. For details
+ * {@see <a href=https://github.com/hiero-ledger/hiero-improvement-proposals/blob/main/HIP/hip-1215.md>HIP-1215</a>}
+ */
+@Singleton
+public class ScheduleCallTranslator extends AbstractCallTranslator<HssCallAttempt> {
+
+    public static final SystemContractMethod SCHEDULED_CALL = SystemContractMethod.declare(
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java]: `SCHEDULE_CALL` (no "D"), same for `SCHEDULED_CALL_WITH_SENDER` below

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+                    "executeCallOnSenderSignature(address,address,uint256,uint256,uint64,bytes)",
+                    ReturnTypes.RESPONSE_CODE_ADDRESS)
+            .withCategories(Category.SCHEDULE);
+
+    @Inject
+    public ScheduleCallTranslator(
+            @NonNull final SystemContractMethodRegistry systemContractMethodRegistry,
+            @NonNull final ContractMetrics contractMetrics) {
+        super(SystemContractMethod.SystemContract.HSS, systemContractMethodRegistry, contractMetrics);
+        registerMethods(SCHEDULED_CALL);
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java]: nitpicking: could be a single call: `registerMethods(SCHEDULE_CALL, SCHEDULE_CALL_WITH_SENDER, EXECUTE_CALL_ON_SENDER_SIGNATURE);`

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        final BigInteger value = call.get(paramIndex++);
+        final var callData = Bytes.wrap((byte[]) call.get(paramIndex));
+
+        // convert parameters
+        final var contractId = ConversionUtils.asContractId(
+                attempt.enhancement().nativeOperations().entityIdFactory(), ConversionUtils.fromHeadlongAddress(to));
+
+        Set<Key> keys = attempt.keySetFor();
+        // create TransactionBody
+        TransactionBody body = TransactionBody.newBuilder()
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java]: nitpicking: this could be final?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+                .transactionID(attempt.enhancement().nativeOperations().getTransactionID())
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java]: Hmh, not sure I understand what's happening here. Isn't `attempt.enhancement().nativeOperations().getTransactionID()` the ID of the transaction currently being executed? If so, does the scheduled transaction get the same ID? What if more calls are scheduled within a single transaction?

Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java]: Ahh I think I see. So this outer `TransactionBody` isn't the body of the scheduled transaction, it's a child call that  will do the actual schedule.

Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java]: Do you think it's worth adding a small comment explaining that? It wasn't obvious to me when I first saw it.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulenative/ScheduleNativeTranslator.java [hunk 14]
```diff
@@ -50,7 +50,7 @@ public ScheduleNativeTranslator(
             @NonNull final SystemContractMethodRegistry systemContractMethodRegistry,
             @NonNull final ContractMetrics contractMetrics) {
         // Dagger
-        super(SystemContractMethod.SystemContract.HAS, systemContractMethodRegistry, contractMetrics);
+        super(SystemContractMethod.SystemContract.HSS, systemContractMethodRegistry, contractMetrics);
```
Comment (lukasz-hashgraph) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulenative/ScheduleNativeTranslator.java:53]: Nice catch 👍

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/ContractsConfig.java (hunk 1 above)
Comment (gkozyryatskyy) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/ContractsConfig.java:68]: true, but functions are different and can be used separately. Like `deleteSchedule` make sense even with out `scheduleCall`. and All `scheduleCall*` function are grouped under `scheduleCall` flag. Same for proxy and non-proxy delete.
I can put `hasScheduleCapacity` under `scheduleCall` because they are related, but I not sure if this make a big sense.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java (hunk 3 above)
Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java:177]: `DispatchType.TOKEN_INFO` wrapped `HederaFunctionality.TOKEN_GET_INFO` wrapped `TokenGetInfo` from `canonical-prices.json`. So it is all related and using similar naming. Renaming TOKEN_INFO->VIEW_CALL or something, will break this 'connection'.

And in out case function `viewGasRequirement` referring to usage logic.

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/ContractsConfig.java (hunk 1 above)
Comment (gkozyryatskyy) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/ContractsConfig.java:68]: Anyway the comment is valid, lets 'vote' on daily.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java (hunk 3 above)
Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java:177]: Anyway the comment is valid, lets 'vote' on daily.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/DispatchForResponseCodeHssCall.java (hunk 4 above)
Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/DispatchForResponseCodeHssCall.java:145]: They all share same address space. so the "number" is unique. it is also can be converted to string, long-zero address, that is unique.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java (hunk 5 above)
Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: this is a typo =) Thank you

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java (hunk 10 above)
Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java]: yeah, also a typo

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/ContractsConfig.java (hunk 1 above)
Comment (stoyanov-st) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/ContractsConfig.java:68]: IIRC we decided to go for a single flag to manage the feature as a whole.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+    public Optional<SystemContractMethod> identifyMethod(@NonNull final HssCallAttempt attempt) {
+        if (attempt.configuration().getConfigData(ContractsConfig.class).systemContractScheduleCallEnabled()) {
+            return attempt.isMethod(SCHEDULE_CALL, SCHEDULE_CALL_WITH_SENDER, EXECUTE_CALL_ON_SENDER_SIGNATURE);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public Call callFrom(@NonNull final HssCallAttempt attempt) {
```
Comment (stoyanov-st) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java:66]: Please follow the convention of wrapping the logic of decoding and preparing the TransactionBody in a Decoder classes.
Same applies for the DeleteScheduleTranslator

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+            return attempt.nativeOperations().entityIdFactory().newScheduleId(number);
+        } else if (attempt.isSelector(DELETE_SCHEDULE_PROXY)) {
+            return attempt.redirectScheduleId();
+        }
+        throw new IllegalStateException("Unexpected function selector");
+    }
+
+    /**
+     * Calculates the gas requirement for a {@code SCHEDULE_CREATE} call.
```
Comment (stoyanov-st) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]:
```suggestion
     * Calculates the gas requirement for a {@code SCHEDULE_DELETE} call.
```

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public Call callFrom(@NonNull final HssCallAttempt attempt) {
+        // create TransactionBody
+        final var body = TransactionBody.newBuilder()
+                // create ScheduleCreateTransactionBody
```
Comment (stoyanov-st) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]:
```suggestion
                // create ScheduleDeleteTransactionBody
```

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/hasschedulecapacity/HasScheduleCapacityTranslator.java [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        // read parameters
+        final var call = HAS_SCHEDULE_CAPACITY.decodeCall(attempt.inputBytes());
+        //        final BigInteger expirySecond = call.get(EXPIRY_SECOND_INDEX);
+        //        final BigInteger gasLimit = call.get(GAS_LIMIT_INDEX);
+        // TODO should be implemented with:
+        //  HIP-1215 https://github.com/hiero-ledger/hiero-improvement-proposals/blob/main/HIP/hip-1215.md
+        //  Issue #20032 https://github.com/hiero-ledger/hiero-consensus-node/issues/20032
+        //  call hasScheduleCapacity when it will be implemented
+
+        return new HasScheduleCapacityCallStub(attempt);
```
Comment (stoyanov-st) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/hasschedulecapacity/HasScheduleCapacityTranslator.java:60]: Can't we just return false here?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/schedule/HasScheduleCapacityTest.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    @BeforeAll
+    public static void setup(final TestLifecycle lifecycle) {
+        lifecycle.doAdhoc(
+                overriding("contracts.systemContract.scheduleService.scheduleCall.enabled", "true"),
+                overriding("contracts.systemContract.scheduleService.deleteSchedule.enabled", "true"),
+                overriding("contracts.systemContract.scheduleService.hasScheduleCapacity.enabled", "true"));
+    }
+
+    @HapiTest
+    @DisplayName("hasScheduleCapacity(uint256,uint256)")
```
Comment (stoyanov-st) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/schedule/HasScheduleCapacityTest.java]: Use more descriptive names that are explaining the test - is it a call, is it positive/negative, is it proxy etc
e.g. `call hasScheduleCapacity() happy path`

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/schedule/HasScheduleCapacityTest.java [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+@HapiTestLifecycle
+public class HasScheduleCapacityTest {
+
+    @Contract(contract = "HIP1215Contract", creationGas = 4_000_000L, isImmutable = true)
+    static SpecContract contract;
+
+    @BeforeAll
+    public static void setup(final TestLifecycle lifecycle) {
+        lifecycle.doAdhoc(
+                overriding("contracts.systemContract.scheduleService.scheduleCall.enabled", "true"),
```
Comment (stoyanov-st) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/schedule/HasScheduleCapacityTest.java]: You might need to restore the configs afterwards when overriding here

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java (hunk 15 above)
Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/schedulecall/ScheduleCallTranslator.java:66]: Added ScheduleCallDecoder
For DeleteScheduleTranslator agreed to use separate method

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/hasschedulecapacity/HasScheduleCapacityTranslator.java (hunk 18 above)
Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/hasschedulecapacity/HasScheduleCapacityTranslator.java:60]: It should return `*Call` and call should return true, because it is not implemented from HSS for now. Discussed and agree to leave stub.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java (hunk 7 above)
Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: The idea here is that `scheduleId` is generated by HSS and should be a long zero. If the user pass random, non long-zero address, it will be rejected with "schedule not found" or something on HSS level

Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: as an example we are applying the same logic at [SignScheduleTranslator](https://github.com/hiero-ledger/hiero-consensus-node/blob/main/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/signschedule/SignScheduleTranslator.java#L173)

Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: Yeah, I think it's fine from correctness point of view, but it's just weird that if there'a schedule of ID 1, both
`0x0000...0001` and `0xaabb...0001` and all others are valid ways to refer to it and would be accepted and successfully resolved to ID 1.

I'd personally prefer if there was only one canonical way - the long zero one - and that it was enforced.

Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: Good point. We cant rollback really (this is a HIP requirement) but I can halt on non long-zero. 
What I worry about, is to have different validation logic on similar functions. Like `sign` will work and `delete` not. 
So there is few options:
1. halt if non long-zero on `delete`
2. halt if non long-zero on `delete` and do the same in `sign`
3. change nothing to keep same logic in `delete` and `sign`

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java (hunk 3 above)
Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java:177]: agree to leave as it is

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java (hunk 7 above)
Comment (lukasz-hashgraph) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: Right, I think validation should be consistent across all of them

Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: So lets choose if we need to change both or leave both. I propose to leave as it is =) But will be nice to ask other opinions

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/schedule/HasScheduleCapacityTest.java (hunk 19 above)
Comment (gkozyryatskyy) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/schedule/HasScheduleCapacityTest.java]: changed to a bit more descriptive names

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java (hunk 7 above)
Comment (stoyanov-st) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: You can just use `ConversionUtils.addressToScheduleID()`
Also we are not encoding the shard/realm to long zero addresses

Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: @stoyanov-st you totally right, but i on purpose not use it to have same logic as at [SignScheduleTranslator](https://github.com/hiero-ledger/hiero-consensus-node/blob/main/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/signschedule/SignScheduleTranslator.java#L173). So we should change both or none of them, right?

Comment (stoyanov-st) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: If my memory does not lie we added `addressToSchedule()` at the same time or after `SignSchedule` but with different PRs so that explains why it's not using it. 
Leaving it to you to decide on whichever to use.

Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: `SignSchedule` have unit test with non long-zero addresses.

As for me we should leave it the same as in `SignSchedule`, but @lukasz-hashgraph suggested to change this and i totally understand his point (im really started from it, before `SignSchedule` tests failed =) ).
So we need a `3rd vote` about it to decide in democracy way =)

Comment (gkozyryatskyy) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hss/deleteschedule/DeleteScheduleTranslator.java]: added `isLongZero` check for `DeleteScheduleTranslator`

**Review by mhess-swl — APPROVED** [review]
Looks good for Execution 👍
